### ISSUE TITLE
how to load files in alternative to using libpd?

### ISSUE THREAD

**Issue by maks** [funkyfourier/spitback#2]
@funkyfourier I've gotten a bit bogged down trying to get libPD binding working with Dart FFI, specifically binding to ALSA, so I realised an easier way for my case of wanting to just run on Linux would be to just send in network messages to a running headless (-nogui) instance of standard pd. 

I found an existing OSC library for Dart so I've started using that as a mechanism for sending/receiving messges from spitback, just by adding a PD patch to spitback that acts as a bridge for OSC messages to PD's internal send/receive so as not to need to modify any of the existing spitback PD files except to add this single file to `main.pd`.

So far I've got the basics working: I can send the `libpdhost-tick` message out via OSC to my external application and receive simple message that can do a [send] to set `bpm` so all good there.

But I'm bit confused as to how to load sample/drum wave files. I had a look at how the `drumcontrolsui` patch works, as I want to avoid trying to send the wav data and just have PD read the files just like the samples do and it looks like doing [send `drum0-filename]` with a path to a wav file should work but it doesn't seem to so I wanted to ask if I'm on the wrong track here?

Here is what I have so far in my "oschandler" patch in case it better helps explains what I'm trying to do (the [route] in this patch is intended for future use to allow me to send different combinations of messages with differing sets of parameters):

![Screenshot from 2021-04-24 13-47-47](https://user-images.githubusercontent.com/71999/115946488-d2cc7900-a504-11eb-83b4-7d392bde2e3f.png)

**Comment by funkyfourier** [comment]
I am not 100% sure what you are doing with regards to OSC, but it seems like you have got that down. As far as I can comprehend it seems like it should work.
With regards to loading wavs in drum slots you should send [_filename_ slot] to the receiver called [_drumplayerid_-filename]. In the case of the default Spitback patch, you should send the list [something.wav 0] to receiver [testdrumplayer-filename], to load something.wav to slot 0 in testdrumplayer.
Do tell if you do not figure it out, then I'll make another attempt at explaining :)

**Comment by funkyfourier** [comment]
Ok, I think I discovered something here. You also have to send the samplerate to the receiver [_drumplayerid_-samplerate-_slot_]. In the following example I have made a drumplayer called "maksdrums". I can then load the wav "../../wavs/drums/kick.wav" to slot 0, and play it by sending "35 127" to "maksdrums-noteon". BUT: I also have to send the sample rate to the receiver [_drumplayerid_-samplerate-_slot_].

![spitback](https://user-images.githubusercontent.com/272493/115972299-cb7d8d80-a54d-11eb-9ae5-ee89ca14f2c7.png)

Sending the sample rate is probably done in the java-code of Sputter when drum-instruments are loaded, and I did not realize this had to be done in Spitback too. I'll see if I can fix that. Please leave the issue open until then. It might also be that it needs the sample rate of the wav-file, in order to play it at the right pitch. I'll look into it.

**Comment by maks** [comment]
@funkyfourier thank you so much for looking at this so quickly! 
That was spot on - it all works well for me now! Thanks! 👍🏻 

I realise now that I didn't properly understand that I needed to add a drumplayer object with a name parameter (eg. `maksdrums`) and also I didn't realise I needed to send a list with the "wav-file-path" **and** a second list item with the slot number - I was only sending the file path. 

And the Readme does mention the need to send a message to `[track-id]-samplerate` to set the sample rate, but I had (incorrectly) assumed I didn't need to do that if I was having PD load the wave file instead of data from the libpd host and the Readme doesn't mention that need to append `-slot` to the end of that receiver name, but of course it makes sense, as the rate is per wav file.

It would be nice to have to worry about sending the samplerate, but for my project this is really easy to set now I understand how to do it, so I'd say this is a very low priority issue from my point of view.

Its very exciting to have my Dart code now playing back samples out of Spitback!
Thank you again!
